[WIP] implement QML timer

AIIX opened the pull request:
#### Description
- Moves all logic for timers to QML / JS if platform is available
- Adds support for displaying multiple timers
- Makes the timer UI responsive adds support for both landscape / portrait modes 

![timer-qml](https://user-images.githubusercontent.com/19663666/100273161-5ef06980-2f82-11eb-8ca2-16108505c89a.gif)

#### Type of PR
If your PR fits more than one category, there is a high chance you should submit more than one PR. Please consider this carefully before opening the PR.
_Either delete those that do not apply, or add an x between the square brackets like so: `- [x]`_
- [ ] Bugfix
- [x] Feature implementation
- [ ] Refactor of code (without functional changes)
- [ ] Documentation improvements
- [ ] Test improvements

#### Testing
Interact with timer skill normally with GUI running

#### Documentation
TBD

#### CLA
- [x] CLA

AIIX committed:
move timer logic to qml

AIIX committed:
remove unused and debug

AIIX committed:
clean qml formating

JarbasAl commented on __init__.py:
```diff
@@ -704,7 +690,8 @@ def validate_duration(string):
         self.enable_intent("handle_mute_timer")
         # Start showing the remaining time on the faceplate
-        self.update_display(None)
+        if self.platform != "mycroft_mark_2":
```
can we remove these platform checks? i can have a remote GUI connected to a mark1 or any other enclosure, just send the gui data and if theres is no GUI it will be ignored

AIIX replied:
I'll move this to use self.gui.connected(), the gui check is needed as for mark-1 it would still use the old pathway to do a timer and for gui qml handles the timer

JarbasAl replied:
i dont think even the connected is needed? in the case of a mark1 it should use both functions calls

- regular faceplate stuff
- gui messages for any remote GUI

the advantage of not using self.gui.connected is that if i connect with the timers already set they will still show up

i expect that a mark1 will control the faceplate + display in the QT GUI (if a remote GUI is connected)

Either way by platform checks what i mean is that we should check for mark1 specifically, not for mark2, using the sample example from other comment, if platform == ovos the GUI behaviour is still wanted, but not the mark1 logic

AIIX replied:
The render paths of timers is very different, qml manages and renders the timers for gui, no update data is sent from skill to qml, we only connect the timers signals to slots in the skill for events like expired timer etc, for qml, not adding a check here would have two unsynced timers running and causing all sorts of issues, so its better to disable one path and use one path exclusively if there is a gui.

JarbasAl replied:
it's totally fine if the mark1 doesnt use the GUI then, but please check for platform==mark1  instead of platform != mark2

JarbasAl commented on __init__.py:
```diff
@@ -704,7 +690,8 @@ def validate_duration(string):
         # Start showing the remaining time on the faceplate
-        self.update_display(None)
+        if self.platform != "mycroft_mark_2":
+            self.update_display(None)
```
instead of passing None here, can we just have a default arg in the update_display function directly? a common pattern for handlers that trigger by bus messages and manually is just passing message=None as default arg

AIIX replied:
If you see the check clearly we don't use self.update_display(...) at all anymore because for gui again qml.handles the timers and updates, the above if wanted changed for mark-1 should be another issue since.

JarbasAl replied:
i get that, what im saying is the check should be

```
if self.platform == "mycroft_mark_1":
    self.update_display()
```

and then
```
def update_display(message=None):
    ....
```
or if its not even triggered by messagebus anymore, then just remove the argument

just feels weird passing the None on the function call

JarbasAl commented on __init__.py:
```diff
@@ -91,8 +91,9 @@ def initialize(self):
         # cancel intent, since there are no timers to cancel yet!
-        self.schedule_repeating_event(self.update_display,
-                                      None, 1, name='ShowTimer')
+        if self.platform != "mycroft_mark_2":
```
if i understand correctly this is just meant for mark1, can we do if platform == mark1  instead? as commented bellow the GUI is not mark2 exclusive, eg, if i have platform == ovos i still want the QT GUI to work properly

we also want the GUI data passed for usage with the [gui_debugger/TUI](https://github.com/JarbasAl/Mycroft-TUI)

AIIX replied:
Will move this to self.gui.connected() check, as there might be more headless devices that want to use the general timer path apart from mark 1

AIIX committed:
check if gui is connected instead of platform

krisgesling commented on __init__.py:
```diff
@@ -944,6 +956,37 @@ def _stop_beep(self):
+            
+        remain_time_in_ms = remaining_time * 1000
+            
+        if timer['name']:
+            timer_name = timer['name']
+        else:
+            if idx is None or idx == 1:
+                timer_name = 'Timer'
+                timer_id = 1
+            else:
+                timer_name = 'Timer ' + str(idx)
+                timer_id = idx
```
In this first conditional path we don't get a `timer_id` set. We could do something like:

```Python
        timer_id = idx or 1
        if timer['name']:
            timer_name = timer['name'].capitalize()
        elif timer_id == 1:
            timer_name = 'Timer'
        else:
            timer_name = 'Timer ' + str(timer_id)
```

krisgesling reviewed (commented):
Hey thanks for this I've been using it a bit lately but finally got back to review in more detail.

There's one undefined variable given a specific path, and we may need to tweak the layout a little when two timers are displayed so that the name and countdown scale down, [see example here](https://user-images.githubusercontent.com/2058458/103964893-0b3d8680-51a4-11eb-8d7c-874f227b2fbe.jpg)

AIIX committed:
check content widths dont exceed boundry, fix hours display, percentage

AIIX commented:
> Hey thanks for this I've been using it a bit lately but finally got back to review in more detail.
> 
> There's one undefined variable given a specific path, and we may need to tweak the layout a little when two timers are displayed so that the name and countdown scale down, [see example here](https://user-images.githubusercontent.com/2058458/103964893-0b3d8680-51a4-11eb-8d7c-874f227b2fbe.jpg)

I couldn't replicate the scaling issue on my wave share 5inch display in horizontal mode using the latest qt image from debos, but I have anyways attempted a fix and changed some calculations for the layouts and pushed a new commit, let me know if it worked since I cant replicate it

krisgesling commented on ui/timer.qml:
```diff
@@ -171,8 +173,14 @@ Mycroft.Delegate {
+                    horizontalAlignment: Text.AlignHCenter
+                    width: timerBackground.width - (Kirigami.Units.largeSpacing + Kirigami.Units.smallSpacing)
+                    height: parent.height * 0.35
+                    font.pixelSize: contentWidth > (width * 0.8) ? timerName.height * 0.4 : timerName.height * 0.6
```
The conditional here doesn't seem to be working for me. "Timer 2" is getting truncated (...)

Seems to work well checking against the parent.width eg 
```
contentWidth > (parent.width * 0.8)
```

Will that work on all platforms?

krisgesling commented on ui/timer.qml:
```diff
@@ -183,9 +191,11 @@ Mycroft.Delegate {
                     color: "white"
                     font.family: "Noto Sans"
-                    font.pixelSize: parent.height * 0.2
+                    font.pixelSize: contentWidth > timerBackground.width ? parent.height * 0.10 : parent.height * 0.18
```
This looked a little small, especially compared to the timer name, can we bump them up a little to 0.15 and 0.25?

krisgesling committed:
Tweak sizing of multi-timer display

krisgesling committed:
Capitalize custom names

also removed additional assignment of `timer_id` 
It is now already set on line 976.

krisgesling commented:
Given how small the suggested changes are I just added them, hope that's alright.

There are a few other visual oddities I've noticed on specific pathways but I'll make issues for them and we can get this merged.

Thanks again!
